src/Microsoft.Data.Analysis/PrimitiveColumnContainer.cs

```diff
@@ -257,31 +257,23 @@ public void Apply<TResult>(Func<T?, TResult?> func, PrimitiveColumnContainer<TRe
         {
             for (int b = 0; b < Buffers.Count; b++)
             {
-                ReadOnlyDataFrameBuffer<T> buffer = Buffers[b];
-                long prevLength = checked(Buffers[0].Length * b);
-                DataFrameBuffer<T> mutableBuffer = DataFrameBuffer<T>.GetMutableBuffer(buffer);
-                Buffers[b] = mutableBuffer;
-                Span<T> span = mutableBuffer.Span;
-                DataFrameBuffer<byte> mutableNullBitMapBuffer = DataFrameBuffer<byte>.GetMutableBuffer(NullBitMapBuffers[b]);
-                NullBitMapBuffers[b] = mutableNullBitMapBuffer;
-                Span<byte> nullBitMapSpan = mutableNullBitMapBuffer.Span;
+                ReadOnlyDataFrameBuffer<T> sourceBuffer = Buffers[b];
+                ReadOnlySpan<byte> sourceNullBitMap = NullBitMapBuffers[b].ReadOnlySpan;
 
                 ReadOnlyDataFrameBuffer<TResult> resultBuffer = resultContainer.Buffers[b];
-                long resultPrevLength = checked(resultContainer.Buffers[0].Length * b);
                 DataFrameBuffer<TResult> resultMutableBuffer = DataFrameBuffer<TResult>.GetMutableBuffer(resultBuffer);
                 resultContainer.Buffers[b] = resultMutableBuffer;
                 Span<TResult> resultSpan = resultMutableBuffer.Span;
                 DataFrameBuffer<byte> resultMutableNullBitMapBuffer = DataFrameBuffer<byte>.GetMutableBuffer(resultContainer.NullBitMapBuffers[b]);
                 resultContainer.NullBitMapBuffers[b] = resultMutableNullBitMapBuffer;
                 Span<byte> resultNullBitMapSpan = resultMutableNullBitMapBuffer.Span;
 
-                for (int i = 0; i < span.Length; i++)
+                for (int i = 0; i < Buffers[b].Length; i++)
                 {
-                    long curIndex = i + prevLength;
-                    bool isValid = IsValid(nullBitMapSpan, i);
-                    TResult? value = func(isValid ? span[i] : default(T?));
+                    bool isValid = IsValid(sourceNullBitMap, i);
+                    TResult? value = func(isValid ? sourceBuffer[i] : default(T?));
                     resultSpan[i] = value.GetValueOrDefault();
-                    SetValidityBit(resultNullBitMapSpan, i, value != null);
+                    resultContainer.SetValidityBit(resultNullBitMapSpan, i, value != null);
```

Comment:
It seems like this has a ton of overlap with the ApplyElementwise.

Apply is almost the same as resultContainer.ApplyElementwise ... just a slightly different func signature?

Reply:
So I dug into this a bunch to try and understand the differences and overlap. After doing that investigation - I like your change. It fixes 2 things.

### Thing 1
This method used to get mutable versions of span and nullBitMap  for the SourceContainer or `this` that we're not changing.

The goal of `Apply` is to start with the value from `this` call `func` with the value or `default` for the type and assign the result to `resultContainer`.  Since it's possible the `func` changed a value from null to not null or vice-versa we have to update the validityBit.

Your change makes it so we don't create mutable versions of source ... since it's not needed. **Nothing should change with source.**

### Thing 2
It also fixes a bug where we were setting the validity bit on the SourceContainer because we were calling `this.SetValidityBit` instead of `resultContainer.SetValidityBit`.